framework/src/executioners/FEProblemSolve.C

```diff
@@ -254,60 +268,174 @@ FEProblemSolve::FEProblemSolve(Executioner & ex)
+T
+FEProblemSolve::getParamFromNonlinearSystemVectorParam(const std::string & param_name,
+                                                       unsigned int index) const
+{
+  const auto & param_vec = getParam<std::vector<T>>(param_name);
+  if (index > _num_nl_systems)
+    paramError(param_name,
+               "Vector parameter is requested at index (" + std::to_string(index) +
+                   ") which is larger than number of nonlinear systems (" +
+                   std::to_string(_num_nl_systems) + ").");
+  if (param_vec.size() == 0)
+    paramError(
+        param_name,
+        "This parameter was passed to a routine which cannot handle empty vector parameters");
+  if (param_vec.size() != 1 && param_vec.size() != _num_nl_systems)
+    paramError(param_name,
+               "Vector parameter size (" + std::to_string(param_vec.size()) +
+                   ") is different than the number of nonlinear systems (" +
+                   std::to_string(_num_nl_systems) + ").");
+
+  // User passed only one parameter, assume it applies to all nonlinear systems
+  if (param_vec.size() == 1)
+    return param_vec[0];
+  else
+    return param_vec[index];
+}
+
 bool
 FEProblemSolve::solve()
 {
-  // This loop is for nonlinear multigrids (developed by Alex)
+  // This should be late enough to retrieve the convergence object.
+  // TODO: Move this to a setup phase, which does not exist for SolveObjects
+  if (isParamValid("multi_system_fixed_point_convergence"))
+    _multi_sys_fp_convergence =
+        &_problem.getConvergence(getParam<ConvergenceName>("multi_system_fixed_point_convergence"));
+
+  // Outer loop for multi-grid convergence
+  bool converged = false;
+  unsigned int num_fp_multisys_iters = 0;
   for (MooseIndex(_num_grid_steps) grid_step = 0; grid_step <= _num_grid_steps; ++grid_step)
   {
-    _problem.solve(_nl.number());
-
-    if (_problem.shouldSolve())
+    // Multi-system fixed point loop
+    // Use a convergence object if provided, if not, use a reasonable default of every nested system
+    // being converged
+    num_fp_multisys_iters = 0;
+    converged = false;
+    while (_multi_sys_fp_convergence
+               ? (_multi_sys_fp_convergence->checkConvergence(num_fp_multisys_iters) ==
+                  Convergence::MooseConvergenceStatus::ITERATING)
+               : !converged)
     {
-      if (_problem.converged(_nl.number()))
-        _console << COLOR_GREEN << " Solve Converged!" << COLOR_DEFAULT << std::endl;
-      else
+      // Loop over each system
+      for (const auto sys : _systems)
       {
-        _console << COLOR_RED << " Solve Did NOT Converge!" << COLOR_DEFAULT << std::endl;
-        return false;
+        const bool is_nonlinear = (dynamic_cast<NonlinearSystemBase *>(sys) != nullptr);
+
+        // Call solve on the problem for that system
+        if (is_nonlinear)
+          _problem.solve(sys->number());
+        else
+          _problem.solveLinearSystem(sys->number());
```

Comment:
`sys->number()` will go from 1...(N_nonlinear + N_linear) whereas both `Problem::solve()` and `Problem::solveLinearSystem` will use numbers from 1...N_nonlinear and 1...N_linear. So this would segfault I think if you use linear systems and nonlinear systems together.

Reply:
The main thrust here is correct, however at least MOOSE nonlinear system numbering (and libmesh system numbering) start at 0

Reply:
Haha, sure then the ranges are:
Nonlinear system numbering: 0...N_NLS-1 and as a solver system it is 0...N_NLS-1
Linear system numbering is: 0...N_LS-1  and as a solver system it is N_NLS...(N_NLS + N_LS - 1)
This feels a little like a bad design from my side, sorry about that.